source/s3_chunk_stream.c

```diff
@@ -0,0 +1,268 @@
+    if (aws_input_stream_get_length(parent_stream->current_stream, &current_stream_length)) {
+        return AWS_OP_ERR;
+    }
+    aws_input_stream_destroy(parent_stream->current_stream);
+
+    struct aws_byte_cursor final_chunk_cursor;
+    /* should I use aws_likely here? */
+    if (current_stream_length > 0) {
+        final_chunk_cursor = aws_byte_cursor_from_string(s_final_chunk);
+    } else {
+        final_chunk_cursor = aws_byte_cursor_from_string(s_empty_chunk);
+    }
+    struct aws_byte_cursor post_trailer_cursor = aws_byte_cursor_from_string(s_post_trailer);
+    struct aws_byte_cursor colon_cursor = aws_byte_cursor_from_string(s_colon);
+
+    struct aws_byte_cursor checksum_result_cursor = aws_byte_cursor_from_buf(&parent_stream->checksum_result);
+    if (parent_stream->checksum_result_output &&
+        aws_byte_buf_init_copy_from_cursor(
+            parent_stream->checksum_result_output, aws_default_allocator(), checksum_result_cursor)) {
+        return AWS_OP_ERR;
+    }
+    if (aws_byte_buf_init(
+            &parent_stream->post_chunk_buffer,
+            aws_default_allocator(),
+            final_chunk_cursor.len + parent_stream->checksum_header_name.len + colon_cursor.len +
+                checksum_result_cursor.len + post_trailer_cursor.len)) {
+        goto error;
+    }
+    if (aws_byte_buf_append(&parent_stream->post_chunk_buffer, &final_chunk_cursor) ||
+        aws_byte_buf_append(&parent_stream->post_chunk_buffer, &parent_stream->checksum_header_name) ||
+        aws_byte_buf_append(&parent_stream->post_chunk_buffer, &colon_cursor) ||
+        aws_byte_buf_append(&parent_stream->post_chunk_buffer, &checksum_result_cursor) ||
+        aws_byte_buf_append(&parent_stream->post_chunk_buffer, &post_trailer_cursor)) {
+        goto error;
+    }
+    struct aws_byte_cursor post_chunk_cursor = aws_byte_cursor_from_buf(&parent_stream->post_chunk_buffer);
+    parent_stream->current_stream = aws_input_stream_new_from_cursor(aws_default_allocator(), &post_chunk_cursor);
+    parent_stream->set_current_stream_fn = s_set_null_stream;
+    return AWS_OP_SUCCESS;
+error:
+    aws_byte_buf_clean_up(parent_stream->checksum_result_output);
+    aws_byte_buf_clean_up(&parent_stream->post_chunk_buffer);
+    return AWS_OP_ERR;
+}
+
+static int s_set_chunk_stream(struct aws_chunk_stream *parent_stream) {
+    aws_input_stream_destroy(parent_stream->current_stream);
+    parent_stream->current_stream = parent_stream->checksum_stream;
+    aws_byte_buf_clean_up(&parent_stream->pre_chunk_buffer);
+    parent_stream->checksum_stream = NULL;
+    parent_stream->set_current_stream_fn = s_set_post_chunk_stream;
+    return AWS_OP_SUCCESS;
+}
+
+static int s_aws_input_chunk_stream_seek(
+    struct aws_input_stream *stream,
+    int64_t offset,
+    enum aws_stream_seek_basis basis) {
+
+    struct aws_chunk_stream *impl = stream->impl;
+    return aws_input_stream_seek(impl->current_stream, offset, basis);
```

Comment:
Just remember that we have `aws_http1_stream_write_chunk`, why not just use that?

Reply:
because, this is not an an http chunk, I wasted a month or two adding trailing headers support to http client, before I realized that brawn used aws-chunked encoding 🤦 which is chunked at the client level.